File: src/engine/world.ts
// world.ts — Environment state (hidden from the agent)

export interface WorldCell {
  hasPit: boolean;
  hasWumpus: boolean;
  hasGold: boolean;
}

export type WorldGrid = WorldCell[][];

export interface Position {
  row: number;
  col: number;
}

export interface Percept {
  breeze: boolean;   // adjacent to pit
  stench: boolean;   // adjacent to wumpus
  glitter: boolean;  // gold in current cell
  bump: boolean;     // walked into wall (unused in grid model)
  scream: boolean;   // wumpus killed (unused here)
}

/** Generate a fresh world grid with randomly placed pits, 1 wumpus, 1 gold. */
export function generateWorld(rows: number, cols: number, numPits: number): WorldGrid {
  // Initialise empty grid
  const grid: WorldGrid = Array.from({ length: rows }, () =>
    Array.from({ length: cols }, (): WorldCell => ({
      hasPit: false,
      hasWumpus: false,
      hasGold: false,
    }))
  );

  // Build a list of all cells except [0,0] (agent start)
  const candidates: Position[] = [];
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (r !== 0 || c !== 0) candidates.push({ row: r, col: c });
    }
  }

  // Shuffle candidates (Fisher-Yates)
  for (let i = candidates.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [candidates[i], candidates[j]] = [candidates[j], candidates[i]];
  }

  // Place pits
  const clampedPits = Math.min(numPits, candidates.length - 2);
  for (let i = 0; i < clampedPits; i++) {
    grid[candidates[i].row][candidates[i].col].hasPit = true;
  }

  // Place wumpus
  grid[candidates[clampedPits].row][candidates[clampedPits].col].hasWumpus = true;

  // Place gold
  grid[candidates[clampedPits + 1].row][candidates[clampedPits + 1].col].hasGold = true;

  return grid;
}

/** Return all valid adjacent cells (4-directional). */
export function getAdjacentCells(pos: Position, rows: number, cols: number): Position[] {
  const { row, col } = pos;
  return [
    { row: row - 1, col },
    { row: row + 1, col },
    { row, col: col - 1 },
    { row, col: col + 1 },
  ].filter(p => p.row >= 0 && p.row < rows && p.col >= 0 && p.col < cols);
}

/** Compute percepts at a given position given the world state. */
export function getPercepts(pos: Position, world: WorldGrid, rows: number, cols: number): Percept {
  const cell = world[pos.row][pos.col];
  const adj = getAdjacentCells(pos, rows, cols);

  return {
    breeze: adj.some(p => world[p.row][p.col].hasPit),
    stench: adj.some(p => world[p.row][p.col].hasWumpus),
    glitter: cell.hasGold,
    bump: false,
    scream: false,
  };
}

/** Serialise a position to a string key, e.g. "2_3". */
export function posKey(pos: Position): string {
  return `${pos.row}_${pos.col}`;
}

/** Parse a "r_c" key back into a Position. */
export function keyToPos(key: string): Position {
  const [r, c] = key.split('_').map(Number);
  return { row: r, col: c };
}

File: src/python_engine/world.py
import random
from dataclasses import dataclass
from typing import List, Tuple

@dataclass
class WorldCell:
    has_pit: bool
    has_wumpus: bool
    has_gold: bool

@dataclass
class Position:
    row: int
    col: int

@dataclass
class Percept:
    breeze: bool
    stench: bool
    glitter: bool
    bump: bool
    scream: bool

WorldGrid = List[List[WorldCell]]

def generate_world(rows: int, cols: int, num_pits: int) -> WorldGrid:
    grid = [[WorldCell(False, False, False) for _ in range(cols)] for _ in range(rows)]
    
    candidates = []
    for r in range(rows):
        for c in range(cols):
            if r != 0 or c != 0:
                candidates.append(Position(row=r, col=c))
                
    random.shuffle(candidates)
    
    clamped_pits = min(num_pits, len(candidates) - 2)
    for i in range(clamped_pits):
        grid[candidates[i].row][candidates[i].col].has_pit = True
        
    grid[candidates[clamped_pits].row][candidates[clamped_pits].col].has_wumpus = True
    grid[candidates[clamped_pits + 1].row][candidates[clamped_pits + 1].col].has_gold = True
    
    return grid

def get_adjacent_cells(pos: Position, rows: int, cols: int) -> List[Position]:
    row, col = pos.row, pos.col
    adj = [
        Position(row - 1, col),
        Position(row + 1, col),
        Position(row, col - 1),
        Position(row, col + 1),
    ]
    return [p for p in adj if 0 <= p.row < rows and 0 <= p.col < cols]

def get_percepts(pos: Position, world: WorldGrid, rows: int, cols: int) -> Percept:
    cell = world[pos.row][pos.col]
    adj = get_adjacent_cells(pos, rows, cols)
    
    breeze = any(world[p.row][p.col].has_pit for p in adj)
    stench = any(world[p.row][p.col].has_wumpus for p in adj)
    glitter = cell.has_gold
    
    return Percept(
        breeze=breeze,
        stench=stench,
        glitter=glitter,
        bump=False,
        scream=False
    )

def pos_key(pos: Position) -> str:
    return f"{pos.row}_{pos.col}"

def key_to_pos(key: str) -> Position:
    r, c = map(int, key.split('_'))
    return Position(row=r, col=c)
